internal/distro/fedora31/distro.go

```diff
@@ -366,48 +364,55 @@ func New() *Fedora31 {
 			"tar",
 			"xz",
 		},
-		arches: map[string]fedora31Arch{
-			"x86_64": {
-				name: "x86_64",
-				bootloaderPackages: []string{
-					"grub2-pc",
-				},
-				buildPackages: []string{
-					"grub2-pc",
-				},
-				imageTypes: map[string]imageType{
-					"ami": amiImgType,
-					"ext4-filesystem": ext4FilesystemType,
-					"partitioned-disk": partitionedDisk,
-					"qcow2": qcow2ImageType,
-					"openstack": openstackImgType,
-					"tar": tarImgType,
-					"vhd": vhdImgType,
-					"vmdk": vmdkImgType,
-				},
-			},
-			"aarch64": {
-				name: "aarch64",
-				bootloaderPackages: []string{
-					"dracut-config-generic",
-					"efibootmgr",
-					"grub2-efi-aa64",
-					"grub2-tools",
-					"shim-aa64",
-				},
-				uefi: true,
-				imageTypes: map[string]imageType{
-					"ami": amiImgType,
-					"ext4-filesystem": ext4FilesystemType,
-					"partitioned-disk": partitionedDisk,
-					"qcow2": qcow2ImageType,
-					"openstack": openstackImgType,
-					"tar": tarImgType,
-					"vhd": vhdImgType,
-				},
-			},
+		arches: map[string]fedora31Arch{},
+	}
+
+	x8664 := fedora31Arch{
+		name: "x86_64",
+		bootloaderPackages: []string{
+			"grub2-pc",
 		},
+		buildPackages: []string{
+			"grub2-pc",
+		},
+		imageTypes: map[string]fedora31ImageType{},
+	}
+	x8664.imageTypes = map[string]fedora31ImageType{
```

Comment:
How about introducing a new method
```golang
func(a *fedora31Arch)addImageType(imageType fedora31ImageType) {
    a.imageTypes[imageType.name] = &fedoraImageType{
        arch: a,
        name: imageType.name,
        /* and so on for the rest of the normal copy */
    }
}
```
I was thinking that if really all but the `arch` pointer is shared, we might have wanted to keep the inner imageType as a separate struct, and just point to it, but I think conceptually we may end up with the imageTypes being really different per-arch, so the fact that for now they are just copies could be an implementation detail.

cc: @larskarlitski

Reply:
One final thought: I think the type of the struct we use to list all the output types i in `New()` should be something just local to that function, and not include an `arch` pointer. In the future, we could imagine `addImageType()` doing more than just coping, and in particular having completely different fields than the structs we use for conveniently writing the definitions down.

Reply:
Alternative solution available here:
https://github.com/msehnout/osbuild-composer/commit/d91846509c33417371d937dea0565f647a485192

Reply:
It's not particularly pleasant to read either, but it might be a bit shorter at least.

Reply:
And also a version with the helpers you proposed:
https://github.com/msehnout/osbuild-composer/commit/4e815057554d76c426f985f972196710d51e027a
But I'm wondering if the version with the wrappers wasn't easier to read. Maybe just rename them to not be so confusing?

Reply:
I like this one: msehnout@357858c. Generally, it's imho best to initialize data structures in one step allowing us to write more immutable code (Go is bad at it, sigh...). However, it's impossible to achieve this in `distro` and friends, because it's basically a tree with reverse links, so it cannot be initialized in one step.

I would change the `addArchList` method a bit:
```golang
func (d *Fedora31) setArchList(arches ...arch) {
	d.arches = map[string]arch{}
	for _, arch := range arches {
		arch.distro = d
		d.arches[arch.name] = arch
	}
}
```

1) rename it to `setArchList`
2) use `...` thingy thing in `arches` argument (I cannot recall the name of `...`)
3) clear the map before adding all the arches

Reply:
Thanks for the suggestion!

I've created one more iteration of the "wrapper struct" idea. It is described in more detail in the commit message:
https://github.com/msehnout/osbuild-composer/commit/20e97b47fc54d6da880034daccb24ee2637210c1

and I'm promise I'm done producing more versions of the same code :D

Reply:
> I like this one: [msehnout/osbuild-composer@357858c](https://github.com/msehnout/osbuild-composer/commit/357858c).

I also like this the most, and I agree with Ondrej's suggestions.

>     1. rename it to `setArchList`

Or even `setArchs()`


Though one part of it will not work, you are changing the passed in structs, but then reusing them later. You need something like:

```go
func (d *Fedora31) setArchs(arches ...arch) {
	d.arches = map[string]arch{}
	for _, arch := range arches {
		d.arches[arch.name] = &arch{
                    distro: d,
                    name: arch.name,
                    /* etc for all the other fields */
		}
	}
}
```

Reply:
Also, I think we should agree on spelling of `arch` plural. Is it `archs` or `arches`? We have both in the current codebase and that's bad. I think I like `arches` more, it's easier to pronounce.

Reply:
I did an poll in the channel, and @major, @bcl and @jrusz all agree: "arches" it is (lots of things to change, in that case).